File: platform-sdk-js/src/process/__tests__/ProcessStartBuilder.test.ts
import { ProcessStartBuilder } from '../ProcessStartBuilder'
import { PlatformClient } from '../../client/PlatformClient'
import { ProcessTracker } from '../types'

const mockClient = {
  post: jest.fn(),
} as unknown as PlatformClient

describe('ProcessStartBuilder', () => {
  beforeEach(() => jest.clearAllMocks())

  it('submit sends processKey in body', async () => {
    const tracker: ProcessTracker = { trackingId: 't-1', statusUrl: '/status/t-1' }
    ;(mockClient.post as jest.Mock).mockResolvedValue(tracker)

    const result = await new ProcessStartBuilder(mockClient, 'order-process').submit()

    expect(result.trackingId).toBe('t-1')
    expect(mockClient.post).toHaveBeenCalledWith(
      '/v1/processes',
      expect.objectContaining({ processKey: 'order-process' }),
    )
  })

  it('builder methods return this for chaining', () => {
    const builder = new ProcessStartBuilder(mockClient, 'key')
    expect(builder.businessKey('bk')).toBe(builder)
    expect(builder.idempotencyKey('ik')).toBe(builder)
    expect(builder.variable('k', 'v')).toBe(builder)
    expect(builder.variables({ a: 1 })).toBe(builder)
    expect(builder.callbackUrl('http://cb')).toBe(builder)
  })

  it('businessKey is included in body when set', async () => {
    ;(mockClient.post as jest.Mock).mockResolvedValue({ trackingId: 't', statusUrl: '/s' })
    await new ProcessStartBuilder(mockClient, 'proc')
      .businessKey('biz-123')
      .submit()
    expect(mockClient.post).toHaveBeenCalledWith('/v1/processes',
      expect.objectContaining({ businessKey: 'biz-123' }))
  })
})

File: platform-sdk-java/platform-sdk-process/src/test/java/com/platform/sdk/process/ProcessStartBuilderTest.java
package com.platform.sdk.process;

import com.platform.sdk.core.http.PlatformHttpClient;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.io.IOException;
import java.util.Map;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.*;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
class ProcessStartBuilderTest {

    @Mock
    private PlatformHttpClient http;

    @Test
    void submit_callsPostWithCorrectPayload() throws IOException {
        ProcessTracker tracker = new ProcessTracker();
        tracker.setTrackingId("t-1");
        when(http.post(eq("/v1/processes"), anyMap(), any())).thenReturn(tracker);

        ProcessTracker result = new ProcessStartBuilder(http, "order-process")
                .businessKey("order-123")
                .variable("amount", 99.99)
                .submit();

        assertThat(result.getTrackingId()).isEqualTo("t-1");
        verify(http).post(eq("/v1/processes"), argThat(body -> {
            Map<?, ?> map = (Map<?, ?>) body;
            return "order-process".equals(map.get("processKey"))
                    && "order-123".equals(map.get("businessKey"));
        }), any());
    }

    @Test
    void builder_chainingReturnsSelf() {
        ProcessStartBuilder builder = new ProcessStartBuilder(http, "key");
        assertThat(builder.businessKey("bk")).isSameAs(builder);
        assertThat(builder.idempotencyKey("ik")).isSameAs(builder);
        assertThat(builder.variable("k", "v")).isSameAs(builder);
        assertThat(builder.variables(Map.of("a", 1))).isSameAs(builder);
        assertThat(builder.callbackUrl("http://cb")).isSameAs(builder);
    }
}
